JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,389 @@
+    // ...
+    render() {
+        return (
+          <div>{this.props.name}</div>
+        );
+      }
+}
+
+ComponentName.propTypes = {
+    id: PropTypes.string.isRequired,
+    name: PropTypes.string,
+};
+
+ComponentName.defaultProps = {
+    name: 'Default',
+};
+```
+
+- you can also declare property types as a **static member** inside a class:
+
+``` javascript
+class ComponentName extends Component {
+    static propTypes = {
+        id: PropTypes.string.isRequired,
+        name: PropTypes.string,
+    }
+    // ...
+    render() {
+        return (
+          <div>{this.props.name}</div>
+        );
+      }
+}
+```
+
+- inside the *render* method, preffer accessing the props using **object destructuring**:
+
+``` javascript
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+
+class ComponentName extends Component {
+    // ...
+    render() {
+        const {id, name} = this.props;
+        // ...
+        return (
+          <div>{name}</div>
+        );
+    }
+}
+```
+
+### Component Local State
+- if you have local state modified by the component, preffer using a single object member
+
+``` javascript
+class ComponentName extends Component {
+    // ...
+    
+    const componentState = {
+        pageTitle: 'Client Work',
+        pageSubtitle: 'See who we work with',
+    }
+    
+    render() {
+        return (
+          <h2>{this.componentState.pageTitle}</h2>
+          <h3>{this.componentState.pageSubtitle}</h3>
+        );
+    }
+}
+```
+
+### Conditional Rendering
+- use ternary operators for conditional rendering
+- return **null** if condition is not met
+
+``` javascript
+render() {
+    // ...
+    return (
+        <AlwaysShownComponent />
+        
+        {
+            isConditionMet
+            ? <ConditionallyShownComponent />
+            : null;
+        }
+    );
+}
+```
+
+### Rendering Arrays of Data
+- use array's **map** method to render items from an array
+
+``` javascript
+render() {
+    // ...
+    return (
+        <ul>
+            {
+                booksArray.map((book) => {
+                    <li>{book.title}></li> 
+                }); 
+            }
+        </ul>
+    );
+}
+```
+
+### Binding
+- bind component's methods inside it's constructor
+
+
+``` javascript
+// not that good
+class ComponentName extends Component {
+    // ...
+    
+    onClickHandler(e) {
+        // ...
+    }
+    
+    render() {
+        return (
+            <ChildComponent
+                onClick={this.onCLickHandler.bind(this)}
+            />
+        );
+    }
+}
+
+// good
+class ComponentName extends Component {
+    // ...
+    constructor(props) {
+        super(props);
+        this.onClickHandler = this.onClickHandler.bind(this);
+    } 
+    
+    onClickHandler(e) {
+        // ...
+    }
+    
+    render() {
+        return (
+            <ChildComponent
+                onClick={this.onCLickHandler}
+            />
+        );
+    }
+}
+
+// good
```

Comment:
Looks like you wanted to write something here... perhaps `@action.bound` (not sure it should be here)